### DISCUSSION TITLE
use UPDI line for serial comunication with my computer

### DISCUSSION THREAD

**Discussion by THARTARUS** [SpenceKonde/megaTinyCore#501] [Q&A] [answered]
i'm new the UPDI programming and i have a doubt.
I want to use an attiny404 or 414 (it depends on Mouser stocks) and use the arduino (or another) serial PRINT library for debugging the MCU for printing some values during my testing.

Can i use the Serial library and the UPDI pin (PA0) for comunication or i need some extra connection and connect  RXD,TXD to TXO,RXI of my sparkfun "Serial Basic Breakout"?

![circuit](https://user-images.githubusercontent.com/43825771/128186285-9cc06689-c830-4923-b197-d388528fac25.PNG)

P.S. i don't want to use OPTIBOOT for memory need.

**Comment by hmeijdam** [comment]
I would recommend to use the TXD/RXD with another Serial to USB adapter. That way you don't need to replug anything and can keep another terminal window open connected to your RXD/TXD pins for debugging, while you upload new versions of your sketch via the UPDI wire.

At least that's how I use it.

I also use low memory parts and therefore have written a very minimal sketch that can just print some debugging info over serial. It consumes about 400 bytes, so I incorporate that in my sketch during development and once done delete it from the final version.

Sketch example is in [this thread](https://github.com/SpenceKonde/megaTinyCore/discussions/494#discussioncomment-1088699)

**Reply by mechatroniks-git** [reply]
I'm not aware of using UPDI for both programming and serial debug.

I use this programmer: https://www.tindie.com/products/leonerd/avr-updi-programmer-with-12v/?pt=ac_prod_search, gives you 2 serial ports, 1 for UPDI and 1 for serial debug.

**Comment by SpenceKonde** [comment, marked as answer]
UPDI programing and serial debug use a differerent pin of the microcontroller (and TX and RX need to not be connected in the way that they are) so no, you cannot use the UPDI line for that. I have attempted, without much success to date to make a serial adapter that would swtich modes based on the states of the modem control pin (IDE serial monitor asserts them, serialupdi does not) but have not managed to assemble any successfuly enough to determine whether there are critical design flaws, or I'm just doing a psis poor job of assembly - no real excuse for why I'm having so much trouble, but I've put multiple ICs on different ones baclwards. lifted traces wjhen tryingm to rework, had one work for a short time then smoke,.It was rather depressing considering that it was a simple board with simple parts on it. and I'm at 0/4 succcessful buiild.

Recommed use a second serial adapter for debug serial. I have been persuing trrying to get an affordable usable chip for a dual serial adapter, but so far no luck. Though a friend says he knows a guy in shenzhen who might be able to score some of the parts I need. The problem is that it's too expensive and bulky yo get a 2 port USB hub chip and use that. with a pair of ch340's

**Reply by MCUdude** [reply]
Didn't we discuss the SerialUPDI hardware earlier this summer? Hardware-wise it's pretty straightforward. The DTR line is always low when in UPDI mode. At least that's how Pymcuprog does it. So all that's needed is a USB to serial adapter with DTR/RTS and an analog switch. I have a pile of MC14053's, so that's what I'd use. You can still get the tiny CH330N/340N chip too, so these two would make a nice, little programmer.
Would you mind sharing the details on what's causing the issue? Did you design a PCB for it?

**Comment by SpenceKonde** [comment]
I have thus far failed to assemble a board that I have been confdent is free of assembly/soldering defects such rthat I can analyze it's failure mode. 
I have had multiple dramatic failures, but have always traced back to dubious sdolder joint, reversed part. 

I do not really have time to work on this when I need to release 2.4.0 after fixing a major bug I just found, without which I can'
rt really release it finish the work I've been hired to do, fix two critical bugs in Dx-core and then release 2.0.0 of ATTinycore before I can dedicate significant time to this. Oh and I need to get an order off to the board house THIS WEEKEND TOO. All of these projects are months late, I am drowning! I do not know how uyou seem to have time to work on these things.

**Reply by MCUdude** [reply]
> I do not know how you seem to have time to work on these things.

Well, I'm not really actively developing anything at the moment. All my Arduino cores seem to be stable without any critical known bugs, so at this point, I'm just responding to issues and looking after the stats really. I always have small side projects like the Event library for instance, but nothing that has a deadline.

**Reply by tricklava** [reply]
I found a solution:
-short the serial debug output to Updi
-configure debug output as week driver (i2c style) by using Pullup as H-driver
-attach an L-sensing interrupt to debug output only during the H-pulses 
-interrupt routine immediately disconnects the debug output when UPTI-diode pulls low

works well for me
Regards


/* UPDI mit Debug-Ausgabe auf dem gleichen seriellen PORT
 * UpDi: ShottkyDiode zw RS232 TXD (Kathode) u RXD (Anode) RXD->AtmegaUPDI
 * Einfach TXD PIN_PB2 gegen UPDI kurzschließen
 * Interupt hält bei UPDI=L das Programm an und 
 * IC kann geflashed werden
 * softserial (C) softSerial_joeymorin 
 */

#include <avr/io.h>
#include <avr/interrupt.h>
#include <util/delay.h>
#include "E:\elektonik\mikrokontroller\AVR\myLib\examples\updiDebugSerial\ports.h"

#define BAUDRATE 115200
#define STOP_BITS 1
#define MICROSECONDS_OVERHEAD_ADJUST 2
#define MICROSECONDS_PER_BIT ((F_CPU / BAUDRATE / 16) - MICROSECONDS_OVERHEAD_ADJUST)

//#define USEUSART0

volatile int stop;
ISR(PORTB_PORT_vect){
    SET_AS_INPUT(B2); // serial pin sofort aus
    CLR_INT(B2); // clear interrupt (writing a "1" clears)
    stop=1;
}

static void USART0_init(uint32_t baud_rate) {
    USART0_CTRLC = USART_CMODE_ASYNCHRONOUS_gc | USART_PMODE_DISABLED_gc | USART_SBMODE_1BIT_gc | USART_CHSIZE_8BIT_gc;
    USART0_BAUD = (uint16_t) (F_CPU * 64 / (16 * (float) baud_rate) + 0.5); 
    USART0_CTRLB |= USART_TXEN_bm;  
}

int debugSerialPrintChar (char c) {
  if(!stop) {
    SET_AS_OUTPUT(B2); //pinConfigure(TXD,OUTPUT|PIN_ISC_DISABLE);
    PORT(B2).PIN2CTRL = 0;
  }  
  #if defined USEUSART0
    USART0.TXDATAL = c;
    while (!(USART0_STATUS & USART_TXCIF_bm)){}; //Wait for Transmit complete Interrupt Flag is set
    USART0_STATUS |= USART_TXCIF_bm;  // Flag resetten
  #else
    //start bit
    SET_L(B2); //TX_PORT &= ~(1<<TXD);
    _delay_us(MICROSECONDS_PER_BIT);

    // data bits
    for (uint8_t bit_mask=0x01; bit_mask; bit_mask<<=1) {
      if (c & bit_mask) SET_H(B2); //TX_PORT |= (1<<TXD);
      else              SET_L(B2); // TX_PORT &= ~(1<<TXD);
      _delay_us(MICROSECONDS_PER_BIT);                // ohne delay 3us 
    }
    // stop bit(s)
    SET_H(B2); //TX_PORT |= (1<<TXD);
    _delay_us(MICROSECONDS_PER_BIT * STOP_BITS);
  #endif
  SET_AS_INPUT(B2);     // pinConfigure(TXD,INPUT_PULLUP|PIN_ISC_FALL);
  PORT(B2).PIN2CTRL = PORT_PULLUPEN_bm | PORT_ISC_FALLING_gc; // INPUT_PULLUP|PIN_ISC_FALL);  
  return c;
}
 
void debugSerialPrintString( char myString[]) {
  short i = 0;
  while (myString[i]) {
    debugSerialPrintChar(myString[i]);
    i++;
  }
}

void debugSerialByte(uint8_t byte) {
              /* Converts a byte to a string of decimal text, sends it */
  debugSerialPrintChar('0' + (byte / 100));                        /* Hundreds */
  debugSerialPrintChar('0' + ((byte / 10) % 10));                      /* Tens */
  debugSerialPrintChar('0' + (byte % 10));                             /* Ones */
}


int main(){
  _PROTECTED_WRITE(CLKCTRL.MCLKCTRLB, !CLKCTRL_PEN_bm); // clear Prescaler to run at 20MHz
  while (!(CLKCTRL.MCLKSTATUS & CLKCTRL_OSC20MS_bm)) {};
  sei();
  #if defined USEUSART0
    USART0_init(BAUDRATE);
  #endif
  
  while(1){
    debugSerialPrintChar('j');debugSerialPrintChar('w'); 
    debugSerialByte(43);
    debugSerialPrintChar('\n');
  debugSerialPrintString("Hello World!");
    _delay_ms(300);
  }
}
